File: src/main/java/arc/arc/store/ItemMatcher.java
package arc.arc.store;

import arc.arc.hooks.HookRegistry;
import de.tr7zw.changeme.nbtapi.NBT;
import lombok.Builder;
import lombok.Data;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;

import java.util.Set;
import java.util.concurrent.atomic.AtomicBoolean;

@Builder
@Data
public class ItemMatcher {

    Material material;
    String materialName;
    Set<String> nbt;
    Boolean hasModelData;
    Boolean isSfItem;

    public boolean matches(ItemStack stack) {
        if (this.material != null && stack.getType() == this.material) return true;
        if (this.nbt != null && !this.nbt.isEmpty()) {
            AtomicBoolean allMatch = new AtomicBoolean(false);
            NBT.get(stack, readableItemNBT -> {
                allMatch.set(nbt.stream().allMatch(readableItemNBT::hasTag));
            });
            if (allMatch.get()) return true;
        }
        if (materialName != null) {
            String mat = stack.getType().name().toLowerCase();
            if (mat.matches(materialName)) return true;
        }
        if (hasModelData != null && stack.getItemMeta() != null) {
            if (stack.getItemMeta().hasCustomModelData() == hasModelData) return true;
        }
        if (isSfItem != null) {
            if (stack.getItemMeta() == null) {

            } else {
                if (HookRegistry.sfHook != null &&
                        HookRegistry.sfHook.isSlimefunItem(stack) == isSfItem) return true;
            }
        }
        return false;
    }

    public static ItemMatcher of(Material material) {
        return ItemMatcher.builder().material(material).build();
    }

    public static ItemMatcher ofNbt(String... strings) {
        return ItemMatcher.builder().nbt(Set.of(strings)).build();
    }

    public static ItemMatcher ofRegex(String regex) {
        return ItemMatcher.builder().materialName(regex).build();
    }

    public static ItemMatcher sfItem(boolean isSfItem) {
        return ItemMatcher.builder().isSfItem(isSfItem).build();
    }

    public static ItemMatcher modelData(boolean hasModelData) {
        return ItemMatcher.builder().hasModelData(hasModelData).build();
    }

}

File: src/main/kotlin/ru/arc/store/ItemMatcher.kt
package ru.arc.store

import de.tr7zw.changeme.nbtapi.NBT
import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import ru.arc.hooks.HookRegistry
import ru.arc.util.hasCustomModelDataSafe

data class ItemMatcher(
    val material: Material? = null,
    val materialName: String? = null,
    val nbt: Set<String>? = null,
    val hasModelData: Boolean? = null,
    val isSfItem: Boolean? = null,
) {
    private val materialPattern: Regex? =
        materialName?.let { expression ->
            require(expression.isNotBlank()) { "Material pattern must not be blank" }
            Regex(expression, RegexOption.IGNORE_CASE)
        }

    fun matches(stack: ItemStack): Boolean {
        if (stack.type == material) return true

        if (!nbt.isNullOrEmpty()) {
            val itemNbt = NBT.readNbt(stack)
            if (matchesCustomNbtTags(itemNbt::hasTag)) return true
        }

        if (materialPattern?.matches(stack.type.name) == true) return true

        if (hasModelData != null && stack.itemMeta?.hasCustomModelDataSafe() == hasModelData) {
            return true
        }

        val slimefunHook = HookRegistry.sfHook
        if (isSfItem != null && slimefunHook != null) {
            if (slimefunHook.isSlimefunItem(stack) == isSfItem) return true
        }

        return false
    }

    internal fun matchesCustomNbtTags(hasTag: (String) -> Boolean): Boolean =
        !nbt.isNullOrEmpty() && nbt.all(hasTag)

    companion object {
        @JvmStatic fun of(material: Material) = ItemMatcher(material = material)
        @JvmStatic fun ofNbt(vararg strings: String) = ItemMatcher(nbt = setOf(*strings))
        @JvmStatic fun ofRegex(regex: String) = ItemMatcher(materialName = regex)
        @JvmStatic fun sfItem(isSfItem: Boolean) = ItemMatcher(isSfItem = isSfItem)
        @JvmStatic fun modelData(hasModelData: Boolean) = ItemMatcher(hasModelData = hasModelData)
    }
}
